### PR TITLE
Enhance/gahm converge

### LINKED ISSUE #75: GAHM-based Perturbation Convergence
Default timelimit for [prep.sbatch](https://github.com/oceanmodeling/ondemand-storm-workflow/blob/main/stormworkflow/slurm/prep.sbatch) is 30 minutes. It timed out for Irma on Hercules. 
I increased the time limit to 2 hours, but it didn't help. This is what I got in `slurm-***.setup.out`

```
+ setup_ensemble --track-file /work2/noaa/nosofs/shared/nhc_hurricanes/irma_2017_aa7c2ce2-eaf9-46ec-8298-9e7562ab9ee4/nhc_track/hurricane-track.dat \
--output-directory /work2/noaa/nosofs/shared/nhc_hurricanes/irma_2017_aa7c2ce2-eaf9-46ec-8298-9e7562ab9ee4/setup/ensemble.dir/ \
--num-perturbations 29 \
--mesh-directory /work2/noaa/nosofs/shared/nhc_hurricanes/irma_2017_aa7c2ce2-eaf9-46ec-8298-9e7562ab9ee4/mesh/ \
--hires-region /work2/noaa/nosofs/shared/nhc_hurricanes/irma_2017_aa7c2ce2-eaf9-46ec-8298-9e7562ab9ee4/mesh/hires \
--sample-from-distribution \
--sample-rule korobov \
--date-range-file /work2/noaa/nosofs/shared/nhc_hurricanes/irma_2017_aa7c2ce2-eaf9-46ec-8298-9e7562ab9ee4/setup/dates.csv \
--nwm-file /work2/noaa/nos-surge/smani/data/nwm/NWM_v2.0_channel_hydrofabric/nwm_v2_0_hydrofabric.gdb \
--tpxo-dir /work2/noaa/nos-surge/smani/data/tpxo \
--variables cross_track along_track radius_of_maximum_winds max_sustained_wind_speed \
--pahm-model gahm irma 2017
/work2/noaa/nos-surge/daneshva/miniconda3/envs/stormworkflow/lib/python3.11/site-packages/pyschism/forcing/hycom/gofs.py:8: 
UserWarning: The seawater library is deprecated! Please use gsw instead.   
  import seawater as sw
can't find the 'adcircpy' module
can't find the 'adcircpy' module
can't find the 'adcircpy' module
/work2/noaa/nos-surge/daneshva/miniconda3/envs/stormworkflow/lib/python3.11/site-packages/stormevents/nhc/track.py:183: 
UserWarning: It is recommended to specify the file_deck and/or advisories when reading from file 
   warnings.warn(
slurmstepd: error: *** JOB 2426246 ON hercules-02-26 CANCELLED AT 2024-08-26T14:31:51 DUE TO TIME LIMIT ***
```

### PR DESCRIPTION
**Pull request by WPringle**
In some instances, the code for computing the isotach velocity `Vr` using the GAHM equation could not converge with the original algorithm.

Changes have been made to add a correction rate when calculating the updated `alpha` as iterations become large. 

https://github.com/oceanmodeling/ondemand-storm-workflow/issues/75#issuecomment-2321137953

### CODE REVIEW

**Comment by WPringle** [comment]
@SorooshMani-NOAA The convergence test case is a good potential idea! But I was thinking it is finalized..

**Comment by SorooshMani-NOAA** [comment]
As we discussed in the meeting, we'll test this update on Laura, Irma and Florence (just the setup part) and if all is good we'll merge the PR. I can also add my own test code to assure convergence in ci tests.

**Comment by SorooshMani-NOAA** [comment]
@WPringle I added a test with timeout failure and now I see that Laura 2020 is failing. Can you please spend some more time on it? 

See: https://github.com/noaa-ocs-modeling/EnsemblePerturbation/blob/enhance/gahm_converge/tests/test_perturb.py

From your env (including newly added `pytest-timeout`) you should be able to run:
```bash
pytest -k perturb_perf tests/
```

to see the failure with all of the latest updates. It doesn't actually time out, because your max iter check fails before that. For the older code (before this branch) this test times out on Irma (as expected).

Please let me know what you think

@FariborzDaneshvar-NOAA I'm already running the workflow with this update on PW. But give this test I think I can already skip that!

**Comment by FariborzDaneshvar-NOAA** [comment]
> @FariborzDaneshvar-NOAA I'm already running the workflow with this update on PW. But give this test I think I can already skip that!
@SorooshMani-NOAA I'm not sure if I understood your comment correctly. Is there anything I need to do with this regard?

**Comment by SorooshMani-NOAA** [comment]
@FariborzDaneshvar-NOAA I just meant that I'm running the workflow with the convergence update for testing. So you don't need to worry about it.

**Comment by WPringle** [comment]
@SorooshMani-NOAA thanks man, I will work on it, I have an idea

**Comment by SorooshMani-NOAA** [comment]
@WPringle just an additional note: I tried running the workflow with these updates for 29 (instead of 39) perturbations and it works for 29 for all storms, however in the test code it fails to converge, which is strange!!

**Comment by WPringle** [comment]
@SorooshMani-NOAA The tests should pass now, I implemented a real bisection method.

**Comment by SorooshMani-NOAA** [comment]
@WPringle thanks for the update. .. it seems that now some of the earlier (file-based) tests fail can you check and see if the difference in results is sensible or not, if it is then we can update the test refs and merge.

**Comment by WPringle** [comment]
@SorooshMani-NOAA Yes, doing that now, mostly seems reasonable, very minor adjustment by 1 n mi for a single entry in most cases. but will double check.

**Comment by WPringle** [comment]
@SorooshMani-NOAA I think it is good to go now.

There is one last thing that is interesting to note:
Sometimes the `isotach radii` can be less than the forecasted `RMW`. In that case I preserve that restriction and only search for the `isotach radii < RMW` rather than `> RMW`. 
It is possible also to make a restriction on `isotach radii` to always be `>RMW`, which would result in modifying the original `isotach radii` in the zero perturbation track. But for now, I think what we have is good as the regressed forecasted `RMW` may not be as accurate as the original forecasted `isotach radii`.

File: tests/data/reference/test_monovariate_besttrack_ensemble/vortex_1_variable_perturbation_2.22 [hunk 1]
```diff
@@ -58,11 +58,11 @@ AL, 06, 2018091518,   , BEST,   0, 336N,  798W,  47,  999, TS,  34, NEQ,  136,
 AL, 06, 2018091600,   , BEST,   0, 336N,  802W,  43,  999, TS,  34, NEQ,  128,  127,  148,  152, 1013,  240, 110,  50,   0,   L,   0,    , 270,   2,   FLORENCE,  22
 AL, 06, 2018091606,   , BEST,   0, 336N,  808W,  38, 1000, TS,  34, NEQ,  126,  124,    0,    0, 1013,  260, 110,  40,   0,   L,   0,    , 270,   3,   FLORENCE,  23
 AL, 06, 2018091612,   , BEST,   0, 336N,  815W,  33, 1003, TS,  34, NEQ,    0,    0,    0,    0, 1013,  280, 140,  40,   0,   L,   0,    , 270,   3,   FLORENCE,  24
-AL, 06, 2018091618,   , BEST,   0, 341N,  821W,  28, 1007, TD,   0,    ,    0,    0,    0,    0, 1013,  300, 140,  40,   0,   L,   0,    , 315,   4,   FLORENCE,  25
```
Comment (SorooshMani-NOAA) [inline tests/data/reference/test_monovariate_besttrack_ensemble/vortex_1_variable_perturbation_2.22:61]: @WPringle thanks for checking, just to double check, can you explain why these values changes from `0` to some large values? Is the new approach avoiding some `nan` values that we previously used to get?

Comment (WPringle) [reply tests/data/reference/test_monovariate_besttrack_ensemble/vortex_1_variable_perturbation_2.22:61]: @SorooshMani-NOAA Yes, the old approach didn't actually converge in some cases so would just get a 0. Now it should always converge to something if it is possible to.

**Pull request merged**